### PR TITLE
Less dynamic memory allocation

### PR DESCRIPTION
**Pull request by jesseli2002**
Dynamic memory allocation rasn't been completely removed (need to convert use of `SubPktPtr` in radio.cpp), to limit the number of changes in a PR and make it easier to review. But the next change will be trivial; the hard part of actually implementing the staticly-allocated memory data structure is complete.

New radio code folder "roar" (for ROcket Agnostic Radio), where the radio data structure goes and any other future radio-support code, including expansion to use multiple queues of different priority.

New class is `roar::Buffer`, which replaces RadioQueue and uses an internal circular buffer. API also supplants the PacketBuffWriter, making it very easy to actually get rid of `std::vector` and dynamic memory allocation.

A limit to the number of subpackets storable is also introduced to prevent dynamic memory allocation.

### CODE REVIEW

**Comment by SilentSys** [comment]
[Remove dynamic memory allocation](https://trello.com/c/l0HM18OF/200-remove-dynamic-memory-allocation)

File: avionics/common/src/radio.cpp [hunk 1]
```diff
@@ -48,14 +47,18 @@
 #endif
 
 namespace {
-constexpr uint64_t GND_STN_ADDR_MSB = 0x0013A200;
-constexpr uint64_t GND_STN_ADDR_LSB = 0x41678FC0;
+constexpr uint64_t kFlaregunAddrMsb = 0x0013A200;
```
Comment (akoen) [inline avionics/common/src/radio.cpp:50]: Is the constant naming convention change intentional?

File: avionics/common/src/radio.cpp [hunk 2]
```diff
@@ -81,19 +84,18 @@ class Radio::RadioMembers {
...
 
   private:
     XBee xbee_;
     XBeeAddress64 gnd_addr_;
     ZBTxRequest tx_packet_;
     ZBRxResponse rx;
 
-    RadioQueue tx_q_; // the [ueue] is silent :)
-    uint8_t payload_[RADIO_MAX_SUBPACKET_SIZE];
+    roar::Buffer tx_q_; // the [ueue] is silent :)
```
Comment (akoen) [inline avionics/common/src/radio.cpp:103]: PR on the english language is severely needed in this case

**Review by akoen — DISMISSED** [review]

**Comment by akoen** [comment]
I don't particularly feel qualified to comment on the effectiveness of this approach, but the style looks good

File: avionics/common/src/radio.cpp (hunk 1 above)
Comment (jesseli2002) [reply avionics/common/src/radio.cpp:50]: yep - all-caps are used by macros, which don't obey namespacing rules => prefix constants with `k` (following Google's naming style) to avoid this problem.

File: avionics/envs/board/include/sensors/GPS.h [hunk 3]
```diff
@@ -71,6 +71,7 @@ PLEASE READ ME!
 /*Includes------------------------------------------------------------*/
 #include <HAL/port_impl.h>
 #include <TinyGPS.h>
+#include <array>
```
Comment (SilentSys) [inline avionics/envs/board/include/sensors/GPS.h]: Why is this needed?

File: avionics/common/src/radio.cpp [hunk 4]
```diff
@@ -48,14 +47,18 @@
...
 constexpr uint32_t RADIO_BAUD_RATE = 921600;
 constexpr uint8_t MAX_PACKETS_PER_RX_LOOP = 8;
-constexpr int MAX_QUEUED_BYTES = 800;
+constexpr int kMaxQueuedBytes = 800;
+// 20 bytes per subpkt on average
+constexpr int kMaxQueuedSubpkts = kMaxQueuedBytes / 20;
 
 } // namespace
 
+constexpr unsigned short kPacketPayloadSpace = 255 - ZB_TX_API_LENGTH;
```
Comment (SilentSys) [inline avionics/common/src/radio.cpp:66]: Might be a good idea to add a comment explaining how this value was chosen and why it is important since we have been burned by this multiple times.

Original comment in radio_queue.h was:
```
/*  Many parts of the XBee library use uint8 for length (e.g. getFrameData).
    Even though the HW likely supports more, this is the theoretical max for
    this XBee library implementation. If we deem this was an implementation
    error, we could PR a fix to the library and then increase this */
```
You dont need to use the exact same one, but anything is better than nothing imho

File: avionics/common/src/roar/buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
+#include <cassert>
```
Comment (SilentSys) [inline avionics/common/src/roar/buffer.cpp:1]: What happens if these asserts are hit running on the board?

File: avionics/common/src/roar/buffer.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+void Buffer::write(uint8_t *dat, int len) {
+    int space_remain = data_cap_ - pos_.index;
+    if (len <= space_remain) {
+        // fits within remaining space
+        std::memcpy(&data_[pos_.index], dat, len);
+    } else {
+        // data wraps to start
+        std::memcpy(&data_[pos_.index], dat, space_remain);
+        std::memcpy(data_.get(), dat + space_remain, len - space_remain);
+    }
```
Comment (SilentSys) [inline avionics/common/src/roar/buffer.cpp:106]: Is this only able to wrap once at a time?

Also does this wrapping result in partially overwritten packets or does addIt take care of that?

**Review by SilentSys — CHANGES_REQUESTED** [review]

File: avionics/common/src/roar/buffer.cpp (hunk 5 above)
Comment (jesseli2002) [reply avionics/common/src/roar/buffer.cpp:1]: There's an option to have asserts compiled out under release mode - so that's the plan. Will make a ticket to ensure that this actually happens.

Alternatively, convert asserts to a log call on assertion fail.

File: avionics/envs/board/include/sensors/GPS.h (hunk 3 above)
Comment (jesseli2002) [reply avionics/envs/board/include/sensors/GPS.h]: It's used later in the file; I suppose <array> must've used to be part of HAL/port_impl.h or something

File: avionics/common/src/roar/buffer.cpp (hunk 6 above)
Comment (jesseli2002) [reply avionics/common/src/roar/buffer.cpp:106]: If it needs to wrap more than once then the subpacket size is larger than the actual buffer size, which violates a key underlying assumption. Although I guess it'd be a good idea to assert that.

Comment (SilentSys) [reply avionics/common/src/roar/buffer.cpp:106]: Yeah at least an assert would be good.

File: avionics/envs/board/include/sensors/GPS.h (hunk 3 above)
Comment (SilentSys) [reply avionics/envs/board/include/sensors/GPS.h]: Wait so which change requires this? I feel like if the change wasn't in GPS that indicates that this include should be placed elsewhere / where it is needed.

Comment (jesseli2002) [reply avionics/envs/board/include/sensors/GPS.h]: In either case it'd be poor practice to leave it out - std::array is used in this header file.
It might've been something from the radio header which included <array> but I'm not sure. But this is where the header is needed, and it's not immediately clear where else.

Comment (SilentSys) [reply avionics/envs/board/include/sensors/GPS.h]: Ah okay I see. Yeah as long as its used here.

**Review by SilentSys — APPROVED** [review]
